homeassistant/components/ring/coordinator.py

```diff
@@ -0,0 +1,102 @@
+        # and start a config flow with SOURCE_REAUTH (async_step_reauth)
+        raise ConfigEntryAuthFailed from err
+    except ring_doorbell.RingTimeout as err:
+        raise UpdateFailed(
+            f"Timeout communicating with API{msg_suffix}: {err}"
+        ) from err
+    except ring_doorbell.RingError as err:
+        raise UpdateFailed(f"Error communicating with API{msg_suffix}: {err}") from err
+
+
+@dataclass
+class RingDeviceData:
+    """RingDeviceData."""
+
+    device: RingGeneric
+    history: Optional[list] = None
+
+
+class RingDataCoordinator(DataUpdateCoordinator[dict[int, RingDeviceData]]):
+    """Base class for device coordinators."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        ring_api,
+    ) -> None:
+        """Initialize my coordinator."""
+        super().__init__(
+            hass,
+            name="devices",
+            logger=_LOGGER,
+            update_interval=SCAN_INTERVAL,
+        )
+        self.ring_api = ring_api
+        self.first_call = True
+
+    async def _async_update_data(self):
+        """Fetch data from API endpoint."""
+        update_method = "update_data" if self.first_call else "update_devices"
+        await _call_api(self.hass, getattr(self.ring_api, update_method))
+        self.first_call = False
+        data = {}
+        devices = self.ring_api.devices()
+        for device_type in devices:
+            for device in devices[device_type]:
+                # Don't update all devices in the ring api, only those that set
+                # their device id as context when they subscribed.
+                if device.id in set(self.async_contexts()):
+                    data[device.id] = RingDeviceData(device=device)
+                    if hasattr(device, "history"):
+                        data[device.id].history = await _call_api(
+                            self.hass,
+                            lambda device: device.history(limit=10),
+                            device,
+                            msg_suffix=f" for device {device.name}",  # device_id is the mac
+                        )
+                    await _call_api(
+                        self.hass,
+                        device.update_health_data,
+                        msg_suffix=f" for device {device.name}",
+                    )
+        return data
+
+
+class RingNotificationsCoordinator(DataUpdateCoordinator[None]):
+    """Global notifications coordinator."""
+
+    def __init__(self, hass: HomeAssistant, ring_api) -> None:
+        """Initialize my coordinator."""
+        super().__init__(
+            hass,
+            logger=_LOGGER,
+            name="active dings",
+            update_interval=NOTIFICATIONS_SCAN_INTERVAL,
+        )
+        self.ring_api = ring_api
+
+    async def _async_update_data(self):
+        """Fetch data from API endpoint."""
+        await _call_api(self.hass, self.ring_api.update_dings)
```

Comment:
This endpoint doesn't return any data, right? Why would you need a `DataUpdateCoordinator` for this? Perhaps you can use `async_track_time_interval` from `homeassistant.helpers.event` to call update_dings on a time interval.

(or maybe I misunderstood what you are trying to achieve here?

Reply:
The endpoint does return data which is stored internally in the api objects `dings_data` variable.  The entity then accesses the data via the `active_alerts()` method which does some filtering based on timestamps/expirey.  It's a bit confusing but it's the logic that was already in place. However the next PR is going to drop this polling coordinator altogether and replace it with realtime push so it doesn't make sense to change this much now.

Reply:
Resolving conversation as @iMicknl gave a thumbs up